### PR TITLE
FIX: bar mixed units, allow ValueError as well

### LINKED ISSUE #13186: ax.bar throws when x axis is pandas datetime
### Bug report

**Bug summary**

ax.bar throws error when x-axis is datetime, behaviour has changed between current master (throws error) and version 3.0.2 (no error)

**Code for reproduction**

```python
import pandas as pd
import matplotlib.pyplot as plt

fig = plt.figure(figsize=(20,10));
ax = fig.gca();

df = pd.DataFrame({'year':[2018,2018,2018],'month':[1,1,1], 'day':[1,2,3], 'value':[1,2,3]})
df['date'] = pd.to_datetime(df[['year', 'month', 'day']])

ax.bar(df['date'], df['value'], width=10,align='center')
```

**Actual outcome**

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-12-500999fd6ea9> in <module>
      8 df['date'] = pd.to_datetime(df[['year', 'month', 'day']])
      9 
---> 10 ax.bar(df['date'], df['value'], width=10,align='center')

~/matplotlib/lib/matplotlib/__init__.py in inner(ax, data, *args, **kwargs)
   1515     def inner(ax, *args, data=None, **kwargs):
   1516         if data is None:
-> 1517             return func(ax, *map(sanitize_sequence, args), **kwargs)
   1518 
   1519         bound = new_sig.bind(ax, *args, **kwargs)

~/matplotlib/lib/matplotlib/axes/_axes.py in bar(self, x, height, width, bottom, align, **kwargs)
   2230             x0 = x
   2231             x = np.asarray(self.convert_xunits(x))
-> 2232             width = self._convert_dx(width, x0, x, self.convert_xunits)
   2233             if xerr is not None:
   2234                 xerr = self._convert_dx(xerr, x0, x, self.convert_xunits)

~/matplotlib/lib/matplotlib/axes/_axes.py in _convert_dx(dx, x0, xconv, convert)
   2055                 dx = [dx]
   2056                 delist = True
-> 2057             dx = [convert(x0 + ddx) - x for ddx in dx]
   2058             if delist:
   2059                 dx = dx[0]

~/matplotlib/lib/matplotlib/axes/_axes.py in <listcomp>(.0)
   2055                 dx = [dx]
   2056                 delist = True
-> 2057             dx = [convert(x0 + ddx) - x for ddx in dx]
   2058             if delist:
   2059                 dx = dx[0]

pandas/_libs/tslibs/timestamps.pyx in pandas._libs.tslibs.timestamps._Timestamp.__add__()

ValueError: Cannot add integral value to Timestamp without freq.
```

**Expected outcome**

**Matplotlib version**
built matplotlib from source

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #13186

Follow up to #12903 where we try and do math first for `bar` (x+dx) before units conversion.  Pandas throws a `ValueError` for a previously working case.  This adds that to the allowable `excepts`...

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
100% agree that pandas datetimes + bar should be tested.  

OTOH, I'm not 100% convinced its matplotlib that should be responsible for those tests if the pandas datetime unit converters are registered, which they are in this case.  See #11664 for where I propose we explictly *unregister* pandas datetime unit conversion so then we can test our own converters (which work with basic pandas objects without the need for their converters).  This was prompted by `import pandas` no longer registering their converters by default.  They backed down on that change, but may re-impliment it at some point.  

Not to say pandas' own converters shouldn't work - but it seems that they should be responsible for reporting breakages to us (like this one).  However, I'm open to the idea we should fully test pandas if that is how it should be done.  

For this particular case, I'm pretty ignorant about how the user got to the point where `bar` worked at all (#13186), so I'm hesitant to add to the tests.  But if someone knoweledgeable of pandas wanted to add or propose a test, I'm more than happy for that to go in...

**Comment by NelleV** [comment]
Have we reported the breakage to pandas? If that is the case, we may want to simply add a test and a comment that this should be fixed in pandas version XXX, and remove the work around and and the test when times come.
Does that seem reasonable ?

**Comment by jklymak** [comment]
Hmm, looking at it again, I guess pandas thinks its a ValueError because `self.freq is None`.   

In any case, not convinced this needs a test from our end, but the change should go in to make the `except` clause here as liberal as possible.  In fact I'd argue its should be a bare except - if adding `convert(x+dx)` fails for any reason give us the old behaviour of just calling `convert(dx)`

**Comment by jklymak** [comment]
... ah apparently `except Exception` is the way to do this cleanly without a dangerous bare exception.  Hopefully that is deemed OK by folks...

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -2057,7 +2057,7 @@ def _convert_dx(dx, x0, xconv, convert):
             dx = [convert(x0 + ddx) - x for ddx in dx]
             if delist:
                 dx = dx[0]
-        except (TypeError, AttributeError) as e:
+        except Exception:
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: please leave a comment justifying (briefly) the nonspecific catch-everything behavior.

Comment (NelleV) [reply lib/matplotlib/axes/_axes.py]: Doesn't that catch everything?

**Comment by NelleV** [comment]
The current solution catches all exceptions, which we should avoid. It means will accidentally catch errors and exception we don't want to catch. I don't think the proposed solution is reasonable (sorry!)

**Comment by NelleV** [comment]
(I need to run, but I'll review this more carefully in the next day or so)

**Comment by jklymak** [comment]
@Nellev I'm happy to go back to explicit exception types, but `except Exceptions` is the pep-8 way to make exception as liberal as possible:

> A bare except: clause will catch SystemExit and KeyboardInterrupt exceptions, making it harder to interrupt a program with Control-C, and can disguise other problems. If you want to catch all exceptions that signal program errors, use except Exception: (bare except is equivalent to except BaseException:).

so I thought this would be OK.  

https://www.python.org/dev/peps/pep-0008/

**Comment by NelleV** [comment]
@jklymak But making catching exception as liberal as possible is not good practice. It means the code "works", in cases where it should probably fail. It actually happened to me several time with Matplotlib that the plot would show and be wrong due to broad error catching such as this one.

I'll dismiss my review, as I am about to board a plane and will be offline for the next 24h, but I think it is better to be as narrow and explicit as possible when catching errors.

**Comment by jklymak** [comment]
Tightened exception 

@dopplershift I can't make the try-except block shorter because the error is sometimes in the procedure that gets the first element.  

Just to be clear - the idea is "try some stuff to get this to work.  If that fails just do what we did before this PR".  So having this try/except block makes things a bit better at no cost (other than code complexity).  

OTOH I don't feel passionately about this solution.  If someone has a better solution for `bar` that would be great.

**Comment by dopplershift** [comment]
@jklymak That's fine--but then for a block that big, I agree with @NelleV that we should be as specific as possible in the exceptions caught. IMHO `except Exception:` is only ok when there's one specific line that you're catching errors from.

**Pull request merged**